File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cpp
using namespace std;
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> counts;
        int maxCount = 0;
        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.size(); right++) {
            counts[s[right]]++;

            maxCount = max(maxCount, counts[s[right]]);

            while ((right - left + 1) - maxCount > k) {
                counts[s[left]]--;
                left++;
            }

            maxLength = max(maxLength, right - left + 1);

        }
        return maxLength;
    }
};

File: Data Structures & Algorithms/permutation-string/submission-1.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        
        counts = [0] * 26
        left = matches = 0

        for i in range(len(s1)):
            counts[ord(s1[i]) - ord('a')] += 1
            counts[ord(s2[i]) - ord('a')] -= 1
        
        for i in range(26):
            if (counts[i] == 0):
                matches += 1
        
        if matches == 26:
            return True
        
        for right in range(len(s1), len(s2)):
            r_idx = ord(s2[right]) - ord('a')
            if (counts[r_idx] == 0):
                matches -= 1
            counts[r_idx] -= 1
            if (counts[r_idx] == 0):
                matches += 1
            
            l_idx = ord(s2[left]) - ord('a')
            if (counts[l_idx] == 0):
                matches -= 1
            counts[l_idx] += 1
            if (counts[l_idx] == 0):
                matches += 1
            left += 1

            if matches == 26:
                return True
            
        return False